torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -392,24 +413,33 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
           std::chrono::milliseconds futureStartTime;
           {
             std::lock_guard<std::mutex> lock{futureMutex_};
-            std::tie(fm, futureStartTime) = futures_[id];
-          }
-          // Not holding lock on markCompleted as this could run callbacks that
-          // call agent_->send
-          fm->markCompleted(std::move(message));
-          {
-            std::lock_guard<std::mutex> lock{futureMutex_};
+            if (futures_.find(id) == futures_.end()) {
+              // Received a completion for a timed out future, drop the recv.
+              // RecvCounts will not be incremented here, it will be incremented
+              // by the sender who has determined the future has timed out.
+              return;
+            }
+
+            const auto& futureInfo = futures_[id];
```

Comment:
we can actually cache the above iterator lookup and use it instead of looking up again